[netxng] Implementation of RRawFileNetXNG.

simonmichal opened the pull request:
# This Pull request:
Adds xroot based implementation of `RRawFile` class, the RRawFileNetXNG.
As the xroot TFile implementation uses the `NetXNG` naming convention for consistency I have adopted the same convention to the `RRawFile` implementation (please let me know if this should be changed).
This is heavily inspired by the `RRawFileDavix` class (@jblomer : thanks a lot for your help).

## Changes or fixes:
Adds the capability to do remote reads with `root`/`roots` to NTuple.

## Checklist:

- [x] tested changes locally, by running the `tutorials/v7/ntuple/ntpl004_dimuon.C` with the `kNTupleFileName` changed to a `root://` URL.

jblomer commented:
@phsft-bot build

jblomer commented on net/netxng/inc/ROOT/RRawFileNetXNG.hxx:
```diff
@@ -0,0 +1,52 @@
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef NET_NETXNG_INC_ROOT_RRAWFILENETXNG_HXX_
```
Let's simplify the header guard to `ROOT_RRawFileNetXNG`

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+{
+   auto st = pImpl->file.Open( fUrl, XrdCl::OpenFlags::Read );
+   if( !st.IsOK() )
+     throw std::runtime_error( "Cannot open '" + fUrl + "', " +
```
Let's wrap the if block in bracelets

simonmichal replied:
Done.

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+
+#include "ROOT/RRawFileNetXNG.hxx"
+
+#include <TError.h>
```
I guess we don't need that include

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+
+std::unique_ptr<ROOT::Internal::RRawFile> ROOT::Internal::RRawFileNetXNG::Clone() const
+{
+   return std::make_unique<RRawFileNetXNG>( fUrl, fOptions );
```
Add `utility` include

simonmichal replied:
AFAIK the `make_unique` utility comes from `#include <memory>`

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+   if( !st.IsOK() )
+     throw std::runtime_error( "Cannot open '" + fUrl + "', " +
+                               st.ToString() + "; " + st.GetErrorMessage() );
+   if( fOptions.fBlockSize < 0 ) fOptions.fBlockSize = kDefaultBlockSize;
```
Split in two lines (for better debugger support)

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+   std::uint32_t btsread = 0;
+   auto st = pImpl->file.Read( offset, nbytes, buffer, btsread );
+   if( !st.IsOK() )
+     throw std::runtime_error( "Cannot read from '" + fUrl + "', " +
```
Wrap block in bracelets

simonmichal replied:
Done.

jblomer commented on net/netxng/src/RRawFileNetXNG.cxx:
```diff
@@ -0,0 +1,107 @@
+   XrdCl::VectorReadInfo *info = nullptr;
+   auto st = pImpl->file.VectorRead( chunks, nullptr, info );
+   if( !st.IsOK() )
+     throw std::runtime_error( "Cannot do vector read from '" + fUrl + "', " +
```
Wrap block in bracelets

jblomer replied:
If `!st.IsOK()`, is it ensured that info remains `nullptr`? Otherwise we'd need to `delete` it in the error path, too.

simonmichal replied:
yes, it is guaranteed that if `!st.IsOK()` then `info` is not allocated

simonmichal replied:
Block is now in braces.

jblomer reviewed (changes requested):
That is great, very concise, many thanks!

I left a few minor comments. I think the style is generally slightly off (wrt to ROOT standard) when it comes to spacing and parenthesis.

One thing I'd like to discuss before merging is how we can unit test the code. If there is a stable, publicly available XRootD service with test data, we can use the RRawFileDavix unit test as a blueprint.  Otherwise, given that we have access to the xrootd library, perhaps we can spawn a simple ad-hoc server in another thread.

simonmichal commented:
@jblomer : thanks a lot for your comments, I will address them shortly!
 
>One thing I'd like to discuss before merging is how we can unit test the code. If there is a stable, publicly available XRootD >service with test data, we can use the RRawFileDavix unit test as a blueprint. Otherwise, given that we have access to the xrootd >library, perhaps we can spawn a simple ad-hoc server in another thread.

Well, I think one nice thing that we could do is to have an XRootD server with an HTTP plugin exporting same files with both root and HTTP protocols. Then we could simply run all the tests you currently have for the `RRawFileDavix` also for `RRawFileNetXNG`. If you have some stress tests we could even use them to benchmark the two protocols.
Could we use `root.cern.ch` to do this? Otherwise, I can create a small testing service, I would need to see what machine I can get.

Axel-Naumann commented:
We can use eospublic for testing xrootd. We have existing tutorials / tests that use it, see e.g. [test/stressIOPlugins.cxx](https://github.com/root-project/root/blob/de7bcdff077be4ec08892c1d8349d9f8e62dd420/test/stressIOPlugins.cxx#L160)

simonmichal commented:
@jblomer : I've just pushed the requested corrections.

jblomer commented:
@simonmichal There is now `root://eospublic.cern.ch//eos/root-eos/xrootd.test`. You can use it to implement unit tests along the line of the `RRawFileDavix` implementation.

simonmichal commented:
@jblomer : thanks a lot, I will cook it up soonish.

simonmichal commented:
@jblomer : I added few tests equivalent to the Davix tests (sorry for the lack creativity) and I also took the liberty and modified  `root://eospublic.cern.ch//eos/root-eos/xrootd.test` so it is in line with `http://root.cern.ch/files/davix.test`.

```
ctest -I 164,165
Test project /home/simonm/git/root-my/buildtmp
    Start 164: gtest-net-davix-test-RRawFileDavix
1/2 Test #164: gtest-net-davix-test-RRawFileDavix .....   Passed    0.20 sec
    Start 165: gtest-net-netxng-test-RRawFileNetXNG
2/2 Test #165: gtest-net-netxng-test-RRawFileNetXNG ...   Passed    0.05 sec
```

However running the full test suite yielded some failures:

```
99% tests passed, 18 tests failed out of 2101

Label Time Summary:
cling                  =  69.10 sec
longtest               = 1541.18 sec
matrix                 =   0.22 sec
multithreaded          = 206.29 sec
python_runtime_deps    =  27.45 sec
regression             =  68.69 sec
roottest               =  69.32 sec
tutorial               = 1559.92 sec

Total Test time (real) = 5594.32 sec

The following tests FAILED:
	  3 - pyunittests-pyroot-dependency-versions (Failed)
	 56 - pyunittests-pyroot-numbadeclare (Failed)
	 57 - test-import-numba (Failed)
	918 - tutorial-dataframe-df026_AsNumpyArrays-py (Failed)
	920 - test-import-pandas (Failed)
	961 - tutorial-pyroot-pyroot004_NumbaDeclare-py (Failed)
	962 - test-import-numba (Failed)
	1219 - roottest-python-JupyROOT-cppcompleter_doctest (Failed)
	1220 - roottest-python-JupyROOT-handlers_doctest (Failed)
	1221 - roottest-python-JupyROOT-utils_doctest (Failed)
	1222 - roottest-python-JupyROOT-importROOT_notebook (Failed)
	1223 - roottest-python-JupyROOT-simpleCppMagic_notebook (Failed)
	1224 - roottest-python-JupyROOT-thread_local_notebook (Failed)
	1225 - roottest-python-JupyROOT-ROOT_kernel_notebook (Failed)
	1226 - roottest-python-JupyROOT-tpython_notebook (Failed)
	1227 - roottest-python-JupyROOT-Cpp_IMT_Canvas_notebook (Failed)
	1444 - roottest-root-html-runMakeIndex (Failed)
	1627 - roottest-root-meta-execTypedefList-auto (Failed)
```

I don't think it's related to the PR but to be on the safe side I'm reporting it.

simonmichal commented:
Cool, thanks for merging :-)
